### ISSUE TITLE
Reuse pool on dependent promises.

### ISSUE THREAD

**Issue by sethtomy** [supercharge/promise-pool#47]
I'm using @supercharge/promise-pool to limit concurrency in my application. I'm attempting to use the same pool to kick off dependent requests but seeing more than my established concurrency going off (e.g. concurrency of 100 is kicking off 200+ requests before ultimately crashing)

My assumption here is that I can re-use the pool to limit concurrency in totality.

```
    const pool = PromisePool
      .withConcurrency(CONCURRENCY);

    await pool
    .for(roots)
    .process(async (root) => {
      await httpRequest(root);
      await pool
        .for(root.children)
        .process(async (child) => {
          await httpRequest(child);
        });
    });

```
Am I using this incorrectly? (Cross posted [here ](https://stackoverflow.com/questions/72545781/supercharge-promise-pool-how-to-run-dependent-promises-in-same-pool) for reference)

**Comment by marcuspoehls** [comment]
@sethtomy Hey, the package is currently not designed to be used for dependencies. Your example shows that you’re starting the pool instance for `roots` which will then create as many tasks as `CONCURRENCY` is defined.

Inside of the `process()` function, you should not reuse the `pool` instance to assign another array and start the processing. Calling `pool.for(childs)` will replace the `roots`:

```js
    const pool = PromisePool
      .withConcurrency(CONCURRENCY);

    await pool
    .for(roots)
    .process(async (root) => {
      await httpRequest(root);
      await pool
        .for(root.children)  // <--     this will replace the `roots` assigned above
        .process(async (child) => {
          await httpRequest(child);
        });
    });

```

What you could do instead is create a new promise pool instance inside the `process` function of your root pool:

```js
const pool = PromisePool.withConcurrency(CONCURRENCY);

await pool
  .for(roots)
  .process(async (root) => {
    await httpRequest(root)
    
    await PromisePool
      .withConcurrency(4)
      .for(root.children)
      .process(async (child) => {
        await httpRequest(child)
      })
  });
```

**Please notice:** starting a promise pool inside the `process` function of a "root" pool increases the concurrency. For example, if the root pool has a concurrency of 5 and each task creates a child promise pool with a concurrency of 10, you’re running a total of 50 tasks simultaneously.

**Comment by st185227** [comment]
@marcuspoehls this is very helpful thank you!

Is this something you'd like to see possible in the package or is the proposed solution the intentional design?

Naïvely I assumed it would work somewhat similarly to rate limiters I'd used in the past. 

Would love to take a crack at implementing if interest. Great package!

**Comment by marcuspoehls** [comment]
@st185227 Hey Seth, I didn’t have the use case to run dependent tasks inside of the pool.

I’m happy if you’re going to implement this functionality. I guess that makes it clear what you’re trying to solve.

Here’s how I understand how you want the promise pool to behave: you want to ensure a limit of X tasks running in parallel. The tasks can also come from items added from within the `process` function of the root pool. The toot pool should support child tasks. The child tasks then eat up a part of the X allowed, overall tasks (concurrency) so that the pool ensures the amount of X tasks in total is not exceeded. Is that correct?

**Comment by sethtomy** [comment]
> Here’s how I understand how you want the promise pool to behave: you want to ensure a limit of X tasks running in parallel. The tasks can also come from items added from within the `process` function of the root pool. The [total] pool should support child tasks. The child tasks then eat up a part of the X allowed, overall tasks (concurrency) so that the pool ensures the amount of X tasks in total is not exceeded. Is that correct?

Exactly. I've used this in the past off some heuristic on client/server load to script out certain work (usually creating test data/running reports).

**Comment by marcuspoehls** [comment]
@sethtomy Ok, I understand. I’m waiting for Seth (@st185227) to propose a solution/idea. You might want to approach this too :)

**Comment by sethtomy** [comment]
@marcuspoehls I just realized I accidentally posted on my previous work account (not sure how that happened). They're both me :)

**Comment by marcuspoehls** [comment]
@sethtomy 😄 ok, thank you for clarifying it. Then I’ll wait for your attempt and merge request 😃

**Comment by sethtomy** [comment]
@marcuspoehls I looked into implementing this morning and was having a bit of trouble deciding how I wanted to represent "subtasks". Through this ended up finding [maxTotalSockets](https://nodejs.org/api/http.html#agentmaxtotalsockets) option on http/https agents. Through this I was able to completely remove ECONNABORTED errors I was seeing at larger concurrency.

I think I'll table implementation for now as this meets my use case quite well.
